<a href="https://colab.research.google.com/github/docum5/Feature_Engineering/blob/main/OHE_variables_with_many_labels.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## One Hot Encoding - variables with many categories


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


***DATA : https://www.kaggle.com/aditya1702/mercedes-benz-data-exploration/data***

In [ ]:
import pandas as pd
import numpy as np



df = pd.read_csv('/content/drive/MyDrive/Feature Engineering/mercedesbenz.csv', usecols=['X1', 'X2','X3','X4','X5','X6'])
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [ ]:
# lihat berapa banyak label yang ada pada setiap label

df.nunique()

X1    27
X2    44
X3     7
X4     4
X5    29
X6    12
dtype: int64

In [ ]:
# lihat berapa banyak kolom yang akan terbentuk jika kita langsung mengaplikasikan one hot encoding

pd.get_dummies(df, drop_first=True).shape

(4209, 117)

***Dapat kita lihat dengan 6 variabel kategori kita berujung dengan 117 variabel baru***


## KDD Cup Orange Challenge

What can we do instead?

http://proceedings.mlr.press/v7/niculescu09/niculescu09.pdf
In the winning solution of the KDD 2009 cup: "Winning the KDD Cup Orange Challenge with Ensemble 

The Team suggested using 10 most frequent labels convert them into dummy variables using onehotencoding

How can we do that in python?

In [ ]:
# let's find the top 10 most frequent categories for the variable X2

df.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: X2, dtype: int64

In [ ]:
# let's make a list with the most frequent categories of the variable

top_10 = [y for y in df.X2.value_counts().sort_values(ascending=False).head(10).index]
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [ ]:
for label in top_10:
    df[label] = np.where(df['X2']==label,1,0)
    
df[['X2']+top_10].head(40)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0
20   m   0   0   0  1   0  0  0  0  0  0
21  as   1   0   0  0   0  0  0  0  0  0
22   a   0   0   0  0   0  0  0  0  0  0
23   m   0   0   0  1   0  0  0  0  0  0
24  ak   0   0   0  0   1  0  0  0  0  0
25   m   0   0   0  1   0  0  0  0  0  0
26  as   1   0   0  0   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28  as   1   0   0  0   0  0  0  0  0  0
29   m   0   0   0  1   0  0  0  0  0  0
30  as   1   0   0  0   0  0  0  0  0  0
31  as   1   0   0  0   0  0  0  0  0  0
32   m   0   0   0  1   0  0  0  0  0  0
33  as   1   0   0  0   0  0  0  0  0  0
34  as   1   0   0  0   0  0  0  0  0  0
35  ai   0   0   1  0   0  0  0  0  0  0
36  as   1   0   0  0   0  0  0  0  0  0
37   k   0   0   0  0   0  0  0  0  0  0
38  as   1   0   0  0   0  0  0  0  0  0
39  as   1   0   0  0   0  0  0  0  0  0

In [ ]:
# get whole set of dummy variables, for all the categorical variables

def one_hot_encoding_top_x(df, variable, top_x_labels):
    # function to create the dummy variables for the most frequent labels
    # we can vary the number of most frequent labels that we encode
    
    for label in top_x_labels:
        df[variable+'_'+label] = np.where(df[variable]==label, 1, 0)



In [ ]:
# read the data again
df = pd.read_csv('/content/drive/MyDrive/Feature Engineering/mercedesbenz.csv', usecols=['X1', 'X2','X3','X4','X5','X6'])

# encode X2 into the 10 most frequent categories
one_hot_encoding_top_x(df, 'X2', top_10)
df.head()

X1  X2 X3 X4 X5 X6  X2_as  ...  X2_m  X2_ak  X2_r  X2_n  X2_s  X2_f  X2_e
0  v  at  a  d  u  j      0  ...     0      0     0     0     0     0     0
1  t  av  e  d  y  l      0  ...     0      0     0     0     0     0     0
2  w   n  c  d  x  j      0  ...     0      0     0     1     0     0     0
3  t   n  f  d  x  l      0  ...     0      0     0     1     0     0     0
4  v   n  f  d  h  d      0  ...     0      0     0     1     0     0     0

[5 rows x 16 columns]